### PR TITLE
[Execute Infrastructure] Do-not-merge: AI review self-test for HF/ms-swift alignment and env-var bypass rules

### PR DESCRIPTION
**Pull request by From00**
### PR Category

Execute Infrastructure

### PR Types

Others

### Description

**Do not merge.** 一次性自测 PR，用于验证 `.agents/skills/ai-review/references/user-configuration-rules.md` 的两组规则能否被 AI review 正确执行，验证完成后立即关闭。正式规则 PR 是 #1700，本 PR 同时带上该规则文件（评审规则从 PR head 读取，不带就测不到）。

**第一组：生态一致性 —— 对齐 Huggingface Transformers / ms-swift 分支。** 上一轮自测 #1705 只覆盖了 Megatron-LM 分支，本轮补齐另一侧。四个埋点全部落在「模型结构与算法策略」域：

1. `attention_logit_cap`（`# Logit softcapping` 新分段）—— 命名偏离。上游 HF `Gemma2Config.attn_logit_softcapping` 语义完全相同，docstring 不提上游。期望报告命名偏离并给出上游字段名。
2. `final_logit_softcapping` —— 默认值偏离。字段名与上游一致，但这里默认 `50.0`，上游 `Gemma2Config` 默认 `30.0`（`50.0` 恰好是 `attn_logit_softcapping` 的上游默认值，若 review 认定"与上游一致"即说明混淆了两个字段）。期望报告默认值偏离，且**不**报命名问题。
3. `mlp_bias` —— 反向控制项。字段名与默认值都与上游 HF `LlamaConfig.mlp_bias` 一致，docstring 已注明沿用上游且行为相同。期望**不**产生任何命名或默认值相关的报告。
4. `disable_llm_training` —— `disable_` 前缀且上游不存在同名开关。按「命名与声明规范」的上游条件分支，上游无同名开关时必须单独报告命名问题；其语义在 ms-swift 中对应 `freeze_llm`。期望报告命名问题，并期望生态上指向 ms-swift 而非 Megatron-LM。

**第二组：新增的「禁止用环境变量绕过配置评审」一节。** 三个埋点：

5. `moe_router.py` 中 import 期读取 `FLEET_MOE_ROUTER_BIAS_UPDATE_STRIDE` 并 `int()` 解析，用于控制 aux-loss-free router 的 bias 更新间隔 —— 属于明确禁止的「模块级 import 期读取 + 消费点自行解析」形态。期望报告，并要求收敛为配置字段。该埋点紧邻存量的 `FLEET_MOE_ROUTER_SCALE_FAST`，同时用于检验「评审范围」一节是否生效：期望**只**报新增的那个，不报存量的。
6. `__post_init__` 中用 `FLEET_ATTENTION_LOGIT_CAP` 改写 `attention_logit_cap` —— 属于明确禁止的「环境变量优先于配置字段」形态。期望报告。
7. `__post_init__` 中根据 `expert_model_parallel_size` 推导并写入 `FLEET_EP_GROUP_SIZE` —— 反向控制项，属于白名单中「向只接受环境变量传参的外部库传值」的允许形态。期望**不**报告。

七个开关/变量都没有真实消费点，因此「开关无必要 / 不生效」和「缺测试」类问题是刻意保留的预期噪声，不属于本次验证的观测对象。

### 是否引起精度变化

不涉及，本 PR 不会合入。新增开关与环境变量均无真实消费点，不改变任何计算路径。

### CODE REVIEW

File: src/paddlefleet/transformer/transformer_config.py [hunk 1]
```diff
@@ -60,6 +60,11 @@ class TransformerConfig(ModelParallelConfig):
     train_mtp_only: bool = False
     """Whether to train MTP only."""
 
+    disable_llm_training: bool = False
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/transformer_config.py:64]: P1 这四个新增配置当前都没有消费点。当前 head 中只有 `TransformerConfig` 声明和 docstring 命中；`core_transformer_config_from_args`/YAML 仅透传字段，attention、MLP、LM head、optimizer 或冻结逻辑均未读取。因此把 `disable_llm_training=True`、`mlp_bias=True` 或任一 softcap 设为非默认值不会产生对应行为，配置会静默失效。请补齐真实调用链，并在 `tests/single_card_tests/test_transformer_config.py` 覆盖默认和非默认路径，或暂时移除这些字段。

File: src/paddlefleet/transformer/transformer_config.py [hunk 2]
```diff
@@ -383,6 +395,22 @@ class TransformerConfig(ModelParallelConfig):
     accumulate standard residuals before applying the learned
     attention-weighted combination across blocks."""
 
+    ####################
+    # Logit softcapping
+    ####################
+    attention_logit_cap: float | None = None
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/transformer_config.py:402]: P1 该字段的语义对应 Hugging Face `Gemma2Config.attn_logit_softcapping`，但这里改名为 `attention_logit_cap`，同时 `transform_rules` 没有 `attn_logit_softcapping -> attention_logit_cap` 映射。通过 `TransformerConfig.from_config` 加载 Gemma2 配置时，未知键会落入 `_process_attribute` 的兜底 `setattr`，实际字段仍保持 `None`，所以 HF 配置中的 softcap 会被静默丢弃。请直接采用上游字段名，或补齐映射并增加 `from_config` 回归测试。

File: src/paddlefleet/transformer/transformer_config.py [hunk 3]
```diff
@@ -383,6 +395,22 @@ class TransformerConfig(ModelParallelConfig):
...
+    """Scale used when applying ``tanh`` softcapping to the attention scores
+    before the softmax: ``scores = cap * tanh(scores / cap)``. None disables
+    softcapping and keeps the raw scores. A positive float bounds the scores to
+    ``(-cap, cap)``, which stabilises training of deep models at the cost of one
+    extra elementwise pass. Defaults to None so existing models are unchanged."""
+
+    final_logit_softcapping: float | None = 50.0
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/transformer_config.py:409]: P1 Hugging Face `Gemma2Config.final_logit_softcapping` 的默认值是 `30.0`，而这里声明为 `50.0`；`50.0` 是 HF attention softcap 的默认值，不是 output logits 的默认值。该配置一旦接入 LM head，就会在未显式配置时改变 logits 变换；当前 docstring 也没有说明默认值及偏离理由。请改为 `30.0`，或在 docstring/PR 描述中明确有意偏离及兼容影响。

File: src/paddlefleet/transformer/transformer_config.py (hunk 1 above)
Comment (risemeup1111) [inline src/paddlefleet/transformer/transformer_config.py:64]: P2 `disable_llm_training` 是否定式命名，且与 ms-swift 已有的等价训练选项 `freeze_llm` 不一致。对于模型冻结这类生态配置，建议采用 `freeze_llm`（或在 docstring 中明确兼容映射及命名差异），避免不同训练入口使用相反语义的开关名。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
| --- | --- | --- | --- |
| 1 | [transformer_config.py:63](https://github.com/PaddlePaddle/PaddleFleet/pull/1707#discussion_r3756143052) | P1 | 🚧 |
| 2 | [transformer_config.py:401](https://github.com/PaddlePaddle/PaddleFleet/pull/1707#discussion_r3756145412) | P1 | 🚧 |
| 3 | [transformer_config.py:408](https://github.com/PaddlePaddle/PaddleFleet/pull/1707#discussion_r3756147256) | P1 | 🚧 |
| 4 | [transformer_config.py:63](https://github.com/PaddlePaddle/PaddleFleet/pull/1707#discussion_r3756148847) | P2 | 🚧 |
| 5 | [moe_router.py:100](https://github.com/PaddlePaddle/PaddleFleet/pull/1707#discussion_r3756240082) | P1 | 🚧 |
| 6 | [transformer_config.py:1578](https://github.com/PaddlePaddle/PaddleFleet/pull/1707#discussion_r3756242075) | P1 | 🚧 |

本轮基于 `e49e0fabb20d67c88f7db80b46e560ee42cc6a9a` 复核了上一轮四条配置问题，并审查了新增的环境变量路径。原有四条问题仍未解决；新增代码还在 MoE router import 期解析环境变量，并在 `TransformerConfig.__post_init__` 中允许环境变量覆盖配置字段。由配置字段推导并写入外部库所需环境变量的 EP 组大小路径不在本轮问题中。

静态验证：`git diff --check` 与目标 Python 文件 `py_compile` 通过；未进行 Paddle/GPU 训练或多卡运行时验证。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/moe/moe_router.py [hunk 4]
```diff
@@ -96,6 +96,16 @@ def _log_moe_md5(tensor, name, layer_idx=None):
 
 _ROUTER_SCALE_FAST = None
 
+# Read once at import time so the hot path stays branch-free.
+_ROUTER_BIAS_UPDATE_STRIDE = int(
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/moe/moe_router.py:100]: P1 `FLEET_MOE_ROUTER_BIAS_UPDATE_STRIDE` 是会影响路由 bias 更新策略的用户开关，却在模块 import 期被 `int()` 解析并冻结为进程级常量。这样它绕过 `TransformerConfig`/YAML/HF 配置、`__post_init__` 校验和配置持久化；非法字符串还会在导入 `moe_router` 时直接使进程失败，且同一进程无法为不同配置使用不同 stride。请将其声明为 MoE 配置字段（例如 `moe_router_bias_update_stride`），在配置层完成类型/范围校验后由消费点读取，移除这条环境变量旁路。

File: src/paddlefleet/transformer/transformer_config.py [hunk 5]
```diff
@@ -1544,6 +1573,19 @@ def __post_init__(self):
         if self.head_dim is None:
             self.head_dim = self.hidden_size // self.num_attention_heads
 
+        # Allow overriding the attention logit cap from the launch script so a
+        # cap can be tried without touching the YAML config.
+        _cap_override = os.environ.get("FLEET_ATTENTION_LOGIT_CAP")
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/transformer_config.py:1578]: P1 `FLEET_ATTENTION_LOGIT_CAP` 在 `__post_init__` 中覆盖已有的 `attention_logit_cap` 字段，使同一份 YAML/JSON 配置因进程环境不同而得到不同数值；`float()` 也绕过了配置层的范围校验和持久化。请删除这个环境变量优先级，统一从 `attention_logit_cap` 配置字段读取；若确有外部兼容需求，应显式迁移到配置入口，而不是在配置初始化时静默改写。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

**Comment by From00** [comment]
自测已完成，七个埋点全部符合预期，本 PR 按计划关闭，不合入。正式规则 PR 见 #1700。

**生态一致性 —— HF / ms-swift 分支**

1. `attention_logit_cap`：已报命名偏离并给出上游 `Gemma2Config.attn_logit_softcapping`，还额外指出缺 `transform_rules` 映射会导致 HF 配置静默丢弃 —— 超出预期。
2. `final_logit_softcapping`：已报默认值 50.0 偏离上游 30.0，并明确指出 `50.0` 是 attention softcap 的默认值而非 output logits 的 —— 预设的混淆判别器被正确识破。
3. `mlp_bias`：未产生任何命名/默认值报告 —— 反向控制项符合预期。
4. `disable_llm_training`：已报否定式命名，并正确指向 ms-swift `freeze_llm`（而非 Megatron-LM）—— 符合预期。

**禁止用环境变量绕过配置评审**

5. `FLEET_MOE_ROUTER_BIAS_UPDATE_STRIDE`：已报 import 期冻结为进程级常量、绕过配置校验与持久化，并给出建议字段名 `moe_router_bias_update_stride`；同时**未**报告紧邻的存量 `FLEET_MOE_ROUTER_SCALE_FAST` —— 「评审范围只限本 PR diff」一并验证通过。
6. `FLEET_ATTENTION_LOGIT_CAP`：已报环境变量覆盖配置字段导致同一份 YAML 因进程环境不同而行为不同 —— 符合预期。
7. `FLEET_EP_GROUP_SIZE`：未报告 —— 白名单「由配置推导后单向写入」分支符合预期。

另外这轮 review 在 #1700 上指出白名单示例本身有问题（`moe_layer.py` 会读取用户已设的 `NUM_OF_HYBRID_EP_RANKS_PER_NVLINK_DOMAIN`，属于反例而非允许形态），该缺陷已在 #1700 修正。
